File: common/src/main/java/dev/kineticcat/complexhex/casting/assemblies/Assemblies.java
package dev.kineticcat.complexhex.casting.assemblies;

import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.phys.Vec3;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static dev.kineticcat.complexhex.Complexhex.id;

@SuppressWarnings("unused")
public class Assemblies {

    public static Map<ResourceLocation, AbstractAssemblyController> ASSEMBLIES = new HashMap<>();

    public static AbstractAssemblyController CUBE = assembly("cube", new SimpleAssemblyController(8, 3));
    public static AbstractAssemblyController TRIANGLE = assembly("triangle", new SimpleAssemblyController(3, 2));

    private static AbstractAssemblyController assembly(String name, AbstractAssemblyController controller) {
        ASSEMBLIES.put(id(name), controller);
        return controller;
    }

    public static AbstractAssemblyController getController(ResourceLocation resLoc) {
        return ASSEMBLIES.get(resLoc);
    }
    public static AbstractAssemblyController getController(String name) {
        return ASSEMBLIES.get(new ResourceLocation(name));
    }
    public static String findController(List<Vec3> verts) {
        for (Map.Entry<ResourceLocation, AbstractAssemblyController> entry : ASSEMBLIES.entrySet()) {
            if (entry.getValue().testVertices(verts)) {
                return entry.getKey().toString();
            }
        }
        return null;
    }


}

File: common/src/main/java/dev/kineticcat/complexhex/casting/assemblies/Assemblies.kt
package dev.kineticcat.complexhex.casting.assemblies

import dev.kineticcat.complexhex.Complexhex
import dev.kineticcat.complexhex.casting.assemblies.complex.AssemblyComplex
import dev.kineticcat.complexhex.casting.assemblies.complex.BelgaerAssemblyComplex
import dev.kineticcat.complexhex.casting.assemblies.complex.HoldoutAssembyComplex
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.phys.Vec3

@Suppress("unused")
class Assemblies {
    companion object {
        var ASSEMBLIES: MutableMap<ResourceLocation, AbstractAssemblyController> = HashMap()

        var CUBE = assembly("cube", IsogonalAssemblyController(8, 3, HoldoutAssembyComplex()))
        var TRIANGLE = assembly("triangle", IsogonalAssemblyController(3, 2, AssemblyComplex()))

        var RHOMBICUBOCTAHEDRON = assembly("rhombicuboctahedron", IsogonalAssemblyController(24, 4, BelgaerAssemblyComplex()))

        private fun assembly(name: String, controller: AbstractAssemblyController): AbstractAssemblyController {
            ASSEMBLIES[Complexhex.id(name)] = controller
            return controller
        }

        fun getController(resLoc: ResourceLocation): AbstractAssemblyController? {
            return ASSEMBLIES[resLoc]
        }

        fun getController(name: String): AbstractAssemblyController? {
            return ASSEMBLIES[ResourceLocation(name)]
        }

        fun findController(verts: List<Vec3>): String? {
            for ((key, value) in ASSEMBLIES) {
                if (value.testVertices(verts)) {
                    return key.toString()
                }
            }
            return null
        }
    }
}
